newrelic-agent/src/main/java/com/newrelic/agent/service/slowtransactions/SlowTransactionService.java

```diff
@@ -0,0 +1,226 @@
+        if (getLogger().isLoggable(Level.FINEST)) {
+            getLogger().finest("Transaction finished with guid " + transactionData.getGuid());
+        }
+        openTransactions.remove(transactionData.getGuid());
+        previouslyReportedTransactions.remove(transactionData.getGuid());
+    }
+
+    // Visible for testing
+    Map<String, Transaction> getOpenTransactions() {
+        return Collections.unmodifiableMap(openTransactions);
+    }
+
+    // Visible for testing
+    Set<String> getPreviouslyReportedTransactions() {
+        return Collections.unmodifiableSet(previouslyReportedTransactions);
+    }
+
+    @Override
+    public void beforeHarvest(String appName, StatsEngine statsEngine) {
+        run();
+    }
+
+    @Override
+    public void afterHarvest(String appName) {
+    }
+
+    // Visible for testing
+    void run() {
+        if (getLogger().isLoggable(Level.FINE)) {
+            getLogger().fine("Identifying slow threads. Open transactions: " + openTransactions.size());
+        }
+        Transaction slowestOpen = null;
+        long slowestOpenMillis = thresholdMillis;
+
+        // Identify the slowest open transaction we haven't yet reported
+        for (Transaction transaction : openTransactions.values()) {
+            // Ignore previously reported transactions
+            if (previouslyReportedTransactions.contains(transaction.getGuid())) {
+                continue;
+            }
+
+            long openMs = System.currentTimeMillis() - transaction.getWallClockStartTimeMs();
+            if (openMs > slowestOpenMillis) {
+                slowestOpen = transaction;
+                slowestOpenMillis = openMs;
+            }
+        }
+
+        if (slowestOpen == null) {
+            getLogger().fine("No new slow transactions identified.");
+            return;
+        }
+
+        // Construct and record SlowTransaction event
+        Map<String, Object> attributes = extractMetadata(slowestOpen, slowestOpenMillis);
+        String guid = slowestOpen.getGuid();
+        if (getLogger().isLoggable(Level.FINE)) {
+            getLogger().fine("Slowest open transaction has guid "
+                    + guid + " has been open for " + slowestOpenMillis + "ms, attributes: " + attributes);
+        }
+        if (insightsService != null) {
+            logger.fine("Sending slow transaction");
+            insightsService.storeEvent(
+                    ServiceFactory.getRPMService().getApplicationName(),
+                    new CustomInsightsEvent(
+                            "SlowTransaction",
+                            System.currentTimeMillis(),
+                            attributes,
+                            DistributedTraceServiceImpl.nextTruncatedFloat()));
+            //insightsService.recordCustomEvent("SlowTransaction", attributes);
+        }
+        addReportedTransaction(guid);
```

Comment:
Instead of tracking `previouslyReportedTransactions`, we could just remove the guid from `openTransactions` right?  That will also cover the case of only reporting a slow transaction once.

Reply:
Thanks - that should be a nice simplification! Will push a commit with the update.